lib/dal/src/func/authoring/save.rs

```diff
@@ -17,175 +13,273 @@ use crate::workspace_snapshot::graph::WorkspaceSnapshotGraphError;
+            invalid => {
+                return Err(FuncAuthoringError::InvalidFuncAssociationsForFunc(
+                    invalid, func.id, func.kind,
+                ))
             }
-        }
-        FuncKind::Authentication => {
-            if let Some(FuncAssociations::Authentication { schema_variant_ids }) = associations {
-                save_auth_func_prototypes(ctx, &func, schema_variant_ids).await?;
+        },
+        FuncKind::Authentication => match associations {
+            FuncAssociations::Authentication { schema_variant_ids } => {
+                update_authentication_associations(ctx, func, schema_variant_ids).await
             }
-        }
-        FuncKind::CodeGeneration => {
-            if let Some(FuncAssociations::CodeGeneration {
+            invalid => {
+                return Err(FuncAuthoringError::InvalidFuncAssociationsForFunc(
+                    invalid, func.id, func.kind,
+                ))
+            }
+        },
+        FuncKind::CodeGeneration => match associations {
+            FuncAssociations::CodeGeneration {
                 schema_variant_ids,
                 component_ids,
                 inputs,
-            }) = associations
-            {
-                save_leaf_prototypes(
+            } => {
+                update_leaf_associations(
                     ctx,
-                    &func,
+                    func,
                     schema_variant_ids,
                     component_ids,
                     &inputs,
                     LeafKind::CodeGeneration,
                 )
-                .await?;
+                .await
             }
-        }
-        FuncKind::Qualification => {
-            if let Some(FuncAssociations::Qualification {
+            invalid => Err(FuncAuthoringError::InvalidFuncAssociationsForFunc(
+                invalid, func.id, func.kind,
+            )),
+        },
+        FuncKind::Qualification => match associations {
+            FuncAssociations::Qualification {
                 schema_variant_ids,
                 component_ids,
                 inputs,
-            }) = associations
-            {
-                save_leaf_prototypes(
+            } => {
+                update_leaf_associations(
                     ctx,
-                    &func,
+                    func,
                     schema_variant_ids,
                     component_ids,
                     &inputs,
                     LeafKind::Qualification,
                 )
-                .await?;
+                .await
             }
-        }
-        func_kind => debug!(%func.id, %func_kind, "no associations for func kind"),
+            invalid => Err(FuncAuthoringError::InvalidFuncAssociationsForFunc(
+                invalid, func.id, func.kind,
+            )),
+        },
+        kind => Err(FuncAuthoringError::FuncCannotHaveAssociations(
+            func.id,
+            kind,
+            associations,
+        )),
```

Comment:
We've beefed up the error checking for associations when saving funcs. Here are two examples.